ompi/mpi/c/finalize.c

```diff
@@ -36,6 +39,65 @@ static const char FUNC_NAME[] = "MPI_Finalize";
 
 int MPI_Finalize(void)
 {
+    /* If --with-spc was specified, print all of the final SPC values
+     * aggregated across the whole MPI run.
+     */
+#if SPC_ENABLE == 1
+    if(!ompi_mpi_spc_dump_enabled)
+        goto skip_dump;
+
+    int i, j, rank, world_size, offset;
+    long long *recv_buffer, *send_buffer;
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+
+    /* Aggregate all of the information on rank 0 using MPI_Gather on MPI_COMM_WORLD */
+    if(rank == 0) {
+        send_buffer = (long long*)malloc(OMPI_NUM_COUNTERS * sizeof(long long));
+        recv_buffer = (long long*)malloc(world_size * OMPI_NUM_COUNTERS * sizeof(long long));
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            send_buffer[i] = events[i].value;
+        }
+        MPI_Gather(send_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG, recv_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG, 0, MPI_COMM_WORLD);
+    } else {
+        send_buffer = (long long*)malloc(OMPI_NUM_COUNTERS * sizeof(long long));
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            send_buffer[i] = events[i].value;
+        }
+        MPI_Gather(send_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG, recv_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG, 0, MPI_COMM_WORLD);
+    }
+
+    /* Once rank 0 has all of the information, print the aggregated counter values for each rank in order */
+    if(rank == 0) {
+        fprintf(stdout, "OMPI Software Counters:\n");
+        offset = 0; /* Offset into the recv_buffer for each rank */
+        for(j = 0; j < world_size; j++) {
+            fprintf(stdout, "World Rank %d:\n", j);
+            for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+                if(attached_event[i]) {
+                    /* If this is a timer-based counter, we need to covert from cycles to usecs */
+                    if(timer_event[i])
+                        SPC_CYCLES_TO_USECS(&recv_buffer[offset+i]);
+                    fprintf(stdout, "%s -> %lld\n", events[i].name, recv_buffer[offset+i]);
+                } else {
+                    fprintf(stdout, "%s -> Disabled\n", events[i].name);
+                }
+            }
+            fprintf(stdout, "\n");
+            offset += OMPI_NUM_COUNTERS;
+        }
+        free(recv_buffer);
+        free(send_buffer);
+        SPC_FINI();
+    } else {
+        free(send_buffer);
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
+ skip_dump:
+#endif
+
```

Comment:
I think this whole thing should be a separate function.

Reply:
like `SPC_Finalize()` or something.

Reply:
I agree, however I'm not sure where to put it.  Should its own .c file in the MPI directory or in the finalize.c file or..?  I can't put it in my driver code.

Reply:
so this is just the shell for MPI_Finalize(). The real thing is in `ompi_mpi_finalize()` call below which is located in `ompi/runtime/ompi_mpi_finalize.c`. You can see they finalize other modules there. Just follow them.